### DISCUSSION TITLE
Unexpected behavior of if statement

### DISCUSSION THREAD

**Discussion by fodisoft** [arendst/Tasmota#22301] [Q&A] [answered]
I don't understand what the problem is! (ESP32)

I made a set of rules and it doesn't always work correctly.
The task would be for each relay to be permanently active for the first half hour after start-up, then switch on every 2 minutes for 10 seconds, for 2 hours, then for 1 hour, also for 10 seconds every 5 minutes.
The problem is that the first and second tasks are carried out, but in the third the relay switches off after 2 second.
The values ​​are lower due to easier control.

Here is the rule set:
```
Rule1
on system#boot do backlog var1 100; var2 2; var3 5; Mem1 35; Mem2 25; Mem3 32; Mem4 30; power1 1; Ruletimer1 60 endon
on Rules#timer=1 do if (%uptime%==%var2%) Var8 60; power1 0; Ruletimer1 60 elseif (%uptime%==%var3%) Var8 120; power1 0; Ruletimer1 60 else Ruletimer1 60 endif endon
on var8#state do Ruletimer2 %var8% endon

Rule2 on Rules#timer=2 do backlog power1 on; delay %var1%; power1 off; Ruletimer2 %var8% endon

Rule3 on DS18B20-1#Temperature>=%Mem1% do Power2 1 endon
on DS18B20-1#Temperature<=%Mem2% do Power2 0 endon
on DS18B20-2#Temperature>=%Mem3% do Power3 1 endon
on DS18B20-1#Temperature<=%Mem4% DO Power3 0 endon
Backlog Rule1 1; Rule2 1; Rule3 1; Rule3 5
```

Can anyone help me why power1 0 is running when it shouldn't be running?

0 replies

**Comment by sfromis** [comment, marked as answer]
Don't use `Delay` that way, it can and will be cut short by other activity.

Instead, you can use `TimedPower1 %var1%00,on` to power on, and automatically off again after var1 * 100 ms. The first argument to `TimedPower` is milliseconds, hence the two zeroes to get from 1/10 seconds to milliseconds. Of course, you can also change the value for `var1` to 10000 for 10 seconds etc.

**Reply by fodisoft** [reply]
I see you have already overtaken me in marking the answer. I'm sorry, I can't be here all the time.

**Comment by fodisoft** [comment]
Thank you very much for your reply, it works perfectly!
I have one more question!
Is it possible to name the variables somehow?
Or how can simple, transparent parameterization be solved?
Maybe an input box for the website?

**Reply by sfromis** [reply]
With rules, there is no variable naming, and only limited ways of doing named parameterization.

You could use commands like `Event maxtemp=24` and rules like `on event#maxtemp do Mem1 %value% endon`, which at the same time would serve as a "reminder" on what the vars are.

With ESP32, you'd have the programming language [Berry](https://tasmota.github.io/docs/Berry/) integrated with Tasmota, with named variables, JSON handling, persistence and ways of defining your own commands. A significant step up from rules, when going above the simplest cases.

If limited to ESP8266, there also exists the feature of [Scripting](https://tasmota.github.io/docs/Scripting-Language/), allowing more programming than rules, but does, well, take some "getting used to".

**Comment by fodisoft** [comment]
Thank you very much for the answer!  

Yes, I've already started dealing with the berry, but I've only just glanced at it.  I haven't read about the scripts on the ESP8266 yet.  Am I right that both can be used to create input boxes, selection lists, checkboxes, radio buttons, and buttons on the website?  So something like a form, the values ​​of which are updated at runtime?
